core-backend/common/src/utils.rs

```diff
@@ -87,25 +58,55 @@ fn smart_truncate(s: &mut String, max_bytes: usize) {
     }
 }
 
-#[derive(Debug, Copy, Clone, derive_more::Display)]
-pub struct LimitedStr<'a>(&'a str);
+/// Wrapped string to fit `core-backend::TRIMMED_MAX_LEN` amount of bytes.
+///
+/// The `Cow` is used to avoid allocating a new `String` when the `LimitedStr` is
+/// created from a `&str`.
+///
+/// Plain `str` is not used because it can't be properly encoded/decoded via scale codec.
+#[derive(
+    TypeInfo, Encode, Decode, Debug, Clone, derive_more::Display, PartialEq, Eq, PartialOrd, Ord,
+)]
+#[codec(crate = scale)]
+pub struct LimitedStr<'a>(Cow<'a, str>);
+
 impl<'a> LimitedStr<'a> {
-    const INIT_ERROR_MSG: &'static str = concat!(
+    const INIT_ERROR_MSG: &str = concat!(
         "String must be less than ",
         stringify!(TRIMMED_MAX_LEN),
         " bytes."
     );
 
-    pub fn new(s: &'a str) -> Result<Self, &'static str> {
+    #[track_caller]
+    pub const fn from_small_str(s: &'a str) -> Self {
+        if s.len() > TRIMMED_MAX_LEN {
+            panic!("{}", Self::INIT_ERROR_MSG)
+        }
+
+        Self(Cow::Borrowed(s))
+    }
+
+    pub fn as_str(&self) -> &str {
+        self.0.as_ref()
+    }
+}
+
+impl<'a> TryFrom<&'a str> for LimitedStr<'a> {
+    type Error = &'static str;
```

Comment:
I suggest
```suggestion
    type Error = LimitedStrTryFromError;
```

because `&'static str` is not convenient to match and work with in general